flexmeasures/ui/crud/assets/services.py

```diff
@@ -0,0 +1,267 @@
+            )
+
+    def validate_on_submit(self):
+        if (
+            hasattr(self, "generic_asset_type_id")
+            and self.generic_asset_type_id.data == -1
+        ):
+            self.generic_asset_type_id.data = (
+                ""  # cannot be coerced to int so will be flagged as invalid input
+            )
+        if hasattr(self, "account_id") and self.account_id.data == -1:
+            del self.account_id  # asset will be public
+        result = super().validate_on_submit()
+        if (
+            hasattr(self, "production_price_sensor_id")
+            and self.production_price_sensor_id is not None
+            and self.production_price_sensor_id.data == -1
+        ):
+            self.production_price_sensor_id.data = None
+        if (
+            hasattr(self, "consumption_price_sensor_id")
+            and self.consumption_price_sensor_id is not None
+            and self.consumption_price_sensor_id.data == -1
+        ):
+            self.consumption_price_sensor_id.data = None
+        return result
+
+    def to_json(self) -> dict:
+        """turn form data into a JSON we can POST to our internal API"""
+        data = copy.copy(self.data)
+        if data.get("longitude") is not None:
+            data["longitude"] = float(data["longitude"])
+        if data.get("latitude") is not None:
+            data["latitude"] = float(data["latitude"])
+
+        if "csrf_token" in data:
+            del data["csrf_token"]
+
+        return data
+
+    def process_api_validation_errors(self, api_response: dict):
+        """Process form errors from the API for the WTForm"""
+        if not isinstance(api_response, dict):
+            return
+        for error_header in ("json", "validation_errors"):
+            if error_header not in api_response:
+                continue
+            for field in list(self._fields.keys()):
+                if field in list(api_response[error_header].keys()):
+                    field_errors = api_response[error_header][field]
+                    if isinstance(field_errors, list):
+                        self._fields[field].errors += api_response[error_header][field]
+                    else:
+                        self._fields[field].errors.append(
+                            api_response[error_header][field]
+                        )
+
+    def with_options(self):
+        if "generic_asset_type_id" in self:
+            self.generic_asset_type_id.choices = [(-1, "--Select type--")] + [
+                (atype.id, atype.name)
+                for atype in db.session.scalars(select(GenericAssetType)).all()
+            ]
+        if "account_id" in self:
+            self.account_id.choices = [(-1, "--Select account--")] + [
+                (account.id, account.name)
+                for account in db.session.scalars(select(Account)).all()
+            ]
+
+    def get_allowed_price_sensor_data(
```

Comment:
Actually, I find that this function and  `get_allowed_inflexible_sensor_data` should better live in `data.services.assets`. (they are not UI-specific)

And then I'd argue to revert the introduction of this `services.py` module, as then our lives become easier when managing our internal stack where we depend on the current module structure.